In [2]:
# !pip install jsonlines

In [16]:
from pydantic import BaseModel, Field
from enum import Enum
from typing import Optional, List
from openai import OpenAI
from typing import Dict
from tqdm import tqdm
import json
import jsonlines
import os
import re
import pandas as pd

from sklearn.metrics import precision_score, recall_score, f1_score, confusion_matrix


In [4]:
client = OpenAI(
    base_url="http://localhost:8001/v1",
    api_key="-",
)
model = client.models.list().data[0].id
model

'/models/Qwen3-0.6B'

In [5]:
class BugStatus(str, Enum):
    BUG = "True"
    NO_BUG = "False"

class ClassificationResult(BaseModel):
    has_bug: BugStatus
    bugs_free_version: str = Field(default=None, description="Optional fix attempt if model detects a bug - Use it only in case of a bug! else, set null")

json_schema = ClassificationResult.model_json_schema()

In [6]:

def benchmark_openai_structured_model(
    test_data: List[Dict],
    client,
    json_schema: dict,
    output_path: str = None,
    temperature: float = 0.0,
    max_tokens: int = 2056,
    save_every: int = 25,
    start_index: int = 0,
):
    benchmark_results = []

    # Check if output file exists and resume from there
    if output_path and os.path.exists(output_path):
        with open(output_path, "r") as f:
            benchmark_results = json.load(f)
        start_index = len(benchmark_results)
        print(f"Resuming from index {start_index}, loaded {len(benchmark_results)} existing results.")

    for i in tqdm(range(start_index, len(test_data)), desc=f"Benchmarking {model}"):
        entry = test_data[i]
        messages = [m for m in entry["messages"] if m["role"] != "assistant"]

        try:
            completion = client.chat.completions.create(
                model=model,
                temperature=temperature,
                max_tokens=max_tokens,
                messages=[
                    entry["messages"][0], # System message
                    entry["messages"][1], # User message
                ],
                # response_format={
                #     "type": "json_schema",
                #     "json_schema": {
                #         "name": "code-diagnostic",
                #         "schema": json_schema
                #     },
                # },
            )

            output_content = completion.choices[0].message.content
            # cleaned_output = re.sub(r"<think>.*?</think>", "", output_content, flags=re.DOTALL)

            benchmark_results.append({
                "index": i,
                "input": messages,
                "output": output_content
            })

        except Exception as e:
            benchmark_results.append({
                "index": i,
                "input": messages,
                "error": str(e)
            })

        # Save progress every `save_every` entries
        if output_path and (i + 1) % save_every == 0:
            with open(output_path, "w") as f:
                json.dump(benchmark_results, f, indent=2)
            

    # Final save
    if output_path:
        with open(output_path, "w") as f:
            json.dump(benchmark_results, f, indent=2)
        print("Final results saved.")

    return benchmark_results

In [7]:
# FIXME - a tmp remove
# os.remove("/home/odedh/SML-For-Debug/benchmark/benchmark_results_Qwen3-0.6B_no_finetunes.json")

with jsonlines.open("/home/odedh/SML-For-Debug/data/dataset/pyresbugs_test.jsonl") as reader:
    test_set = list(reader)

results = benchmark_openai_structured_model(
    test_data=test_set,
    client=client,
    json_schema=json_schema,
    output_path="benchmark_results_Qwen3-0.6B_no_finetunes.json",
    save_every=10,
)

Resuming from index 1002, loaded 1002 existing results.


Benchmarking /models/Qwen3-0.6B: 0it [00:00, ?it/s]

Final results saved.


### Creating the benchmark DF

In [9]:
benchmark_path = "/home/odedh/SML-For-Debug/benchmark/benchmark_results_Qwen3-0.6B_no_finetunes.json"
dataset_path = "/home/odedh/SML-For-Debug/data/dataset/pyresbugs_test.jsonl"

In [10]:
def robust_json_parse(text):
    """
    Attempts to extract and parse a JSON object from a string that may contain
    extra markdown, prefix/suffix, or formatting artifacts.
    """
    if not isinstance(text, str):
        return None

    # Remove Markdown-style code blocks
    cleaned = re.sub(r"```json|```", "", text.strip(), flags=re.IGNORECASE)
    
    # Extract the thinking part if it exists
    thinking_part = re.search(r"<think>(.*?)</think>", cleaned, flags=re.DOTALL)
    if thinking_part:
        cleaned = re.sub(r"<think>.*?</think>", "", cleaned, flags=re.DOTALL)
        thinking_part = thinking_part.group(1).strip()

    # Find the first JSON-like object using regex (handles nested braces)
    match = re.search(r"\{.*\}", cleaned, flags=re.DOTALL)
    if match:
        json_str = match.group(0)
        try:
            json_obj = json.loads(json_str)
            json_obj['thinking'] = thinking_part if thinking_part else None
            return json_obj
        except json.JSONDecodeError:
            # Try to fix common issues like escaped quotes
            try:
                fixed_json_str = json_str.replace("\\'", "'").replace('\\"', '"')
                json_obj = json.loads(fixed_json_str)
                json_obj['thinking'] = thinking_part if thinking_part else None
                return json_obj
            except Exception:
                return None
    return None

In [ ]:
with open(benchmark_path, "r") as f:
    benchmark_data = json.load(f)

# Load JSONL dataset
dataset = []
with open(dataset_path, "r") as f:
    for line in f:
        dataset.append(json.loads(line))

# Build dataframe
records = []

for entry_gt, entry_model in zip(dataset, benchmark_data):
    # Extract ground truth
    user_msg = next(m for m in entry_gt["messages"] if m["role"] == "user")
    assistant_msg = next(m for m in entry_gt["messages"] if m["role"] == "assistant")
    source_code = str(user_msg["content"])
    prompt = entry_gt["messages"][0]["content"]

    try:
        gt_response = json.loads(assistant_msg["content"])
        gt_is_buggy = "BUG FREE" not in gt_response.get("Classification", "").upper()
        gt_fix = gt_response.get("Code", None)
        gt_acronym = gt_response.get("Fault_Acronym", None)
    except Exception:
        gt_is_buggy = None
        gt_fix = None
        gt_acronym = None

    # Extract model output
    try:
        model_output = robust_json_parse(entry_model["output"])
        model_is_buggy = model_output.get("has_bug", "").lower() == "true"
        model_fix = model_output.get("bugs_free_version", None)
        model_thinking = model_output.get("thinking", None)
    except Exception as e:
        model_is_buggy = None
        model_fix = None
        model_thinking = None

    records.append({
        "input_prompt": prompt.strip(),
        "source_code": source_code.strip(),
        "ground_truth_is_buggy": gt_is_buggy,
        "ground_truth_fix": gt_fix,
        "ground_truth_acronym": gt_acronym,
        "model_is_buggy": model_is_buggy,
        "model_suggested_fix": model_fix,
        "model_thinking": model_thinking,
    })

df = pd.DataFrame(records)
df


input_prompt  \
0     You are a code analysis model. Your task is to...   
1     You are a code analysis model. Your task is to...   
2     You are a code analysis model. Your task is to...   
3     You are a code analysis model. Your task is to...   
4     You are a code analysis model. Your task is to...   
...                                                 ...   
997   You are a code analysis model. Your task is to...   
998   You are a code analysis model. Your task is to...   
999   You are a code analysis model. Your task is to...   
1000  You are a code analysis model. Your task is to...   
1001  You are a code analysis model. Your task is to...   

                                            source_code  \
0     def __init__(\n        self,\n        database...   
1     def uuid3(namespace, name):\n    from hashlib ...   
2     def cascaded_union(self):\n        "Returns a ...   
3     def _copy(master_fd, master_read=_read, stdin_...   
4     def __str__(self):\n        template = '(NOT (...   
...                                                 ...   
997   def strip_entities(value):\n    "Returns the g...   
998   def _safe_ParseResult(parts, encoding='utf8', ...   
999   def renew_from_kt(principal, keytab):\n    ren...   
1000  def _get_field(self, field_name: str):\n      ...   
1001  def replace(self, to_replace, value=None, meth...   

      ground_truth_is_buggy  \
0                     False   
1                      True   
2                     False   
3                     False   
4                     False   
...                     ...   
997                   False   
998                   False   
999                   False   
1000                  False   
1001                   True   

                                       ground_truth_fix ground_truth_acronym  \
0                                                  None                 None   
1     def uuid3(namespace, name):\n    from hashlib ...                 WFCD   
2                                                  None                 None   
3                                                  None                 None   
4                                                  None                 None   
...                                                 ...                  ...   
997                                                None                 None   
998                                                None                 None   
999                                                None                 None   
1000                                               None                 None   
1001  def replace(self, to_replace=None, value=None,...                 MLPL   

     model_is_buggy                                model_suggested_fix  \
0              None                                               None   
1             False  def uuid3(namespace, name):\n    from hashlib ...   
2              True  def cascaded_union(self):\n        return GEOS...   
3             False  def _copy(master_fd, master_read= _read, stdin...   
4              True  def __str__(self):\n        template = '(NOT (...   
...             ...                                                ...   
997            None                                               None   
998           False  def _safe_ParseResult(parts, encoding='utf8', ...   
999           False  def renew_from_kt(principal, keytab):\n    ren...   
1000          False  def _get_field(self, field_name: str):\n      ...   
1001          False  def replace(self, to_replace, value=None, meth...   

                                         model_thinking  
0                                                  None  
1     Okay, let's look at this code. The function is...  
2     Okay, let's look at this code. The function is...  
3     Okay, let's look at this code. The function is...  
4     Okay, let's look at this code. The function __...  
...                                                 ..

### EDA

In [17]:
# Filter rows where model prediction exists and ground truth is defined
valid_df = df[df["model_is_buggy"].notnull() & df["ground_truth_is_buggy"].notnull()]

# Ensure both are booleans
y_true = valid_df["ground_truth_is_buggy"].astype(bool)
y_pred = valid_df["model_is_buggy"].astype(bool)

# Basic classification metrics
precision = precision_score(y_true, y_pred)
recall = recall_score(y_true, y_pred)
f1 = f1_score(y_true, y_pred)
conf_matrix = confusion_matrix(y_true, y_pred)

# Per-class F1 score by acronym
f1_per_acronym = {}
acronym_df = valid_df[valid_df["ground_truth_is_buggy"] == True]
for acronym in acronym_df["ground_truth_acronym"].dropna().unique():
    sub_df = acronym_df[acronym_df["ground_truth_acronym"] == acronym]
    if not sub_df.empty:
        y_true_acr = sub_df["ground_truth_is_buggy"].astype(bool)
        y_pred_acr = sub_df["model_is_buggy"].astype(bool)
        score = f1_score(y_true_acr, y_pred_acr, zero_division=0)
        f1_per_acronym[acronym] = score

# Convert to DataFrame for display
f1_acronym_df = pd.DataFrame(list(f1_per_acronym.items()), columns=["Fault_Acronym", "F1_Score"]).sort_values(by="F1_Score", ascending=False)

# Return all to user
{
    "precision": precision,
    "recall": recall,
    "f1_score": f1,
    "confusion_matrix": conf_matrix.tolist(),
    "total_samples": len(valid_df)
}


{'precision': 0.5333333333333333,
 'recall': 0.24060150375939848,
 'f1_score': 0.3316062176165803,
 'confusion_matrix': [[336, 84], [303, 96]],
 'total_samples': 819}

## 🧪 Model Performance Summary (Bug Classification)

The model was evaluated on 819 samples from a structured bug detection benchmark. It was tasked with classifying whether a given code snippet is buggy or bug-free. Below is a summary of its performance:

### 📈 Classification Metrics

- **Precision**: `53.3%`  
  The model correctly identified bugs in 53.3% of the cases it predicted as buggy.

- **Recall**: `24.1%`  
  The model only caught 24.1% of all actual buggy code samples, indicating a high miss rate.

- **F1 Score**: `33.2%`  
  A low F1 score reflects a poor balance between precision and recall, with many false negatives.

- **Total Evaluated Samples**: `819`

---

### 🔍 Confusion Matrix

|                       | Predicted: Bug-Free | Predicted: Buggy |
|-----------------------|---------------------|------------------|
| **Actual: Bug-Free**  | 336 (True Negatives) | 84 (False Positives) |
| **Actual: Buggy**     | 303 (False Negatives) | 96 (True Positives) |

---

### 📊 Key Observations

- The model **performs better at identifying clean code** than detecting buggy code.
- It **misses the majority of actual bugs**, with 303 false negatives.
- While **precision is moderate**, recall is notably low — indicating that the model is overly conservative.
- This behavior suggests that relying solely on this model for bug detection may result in many bugs going undetected.

